crosscluster/ldrrandgen: TestGenerateLDRTable failed

crosscluster/ldrrandgen.TestGenerateLDRTable [failed](https://mesolite.cluster.engflow.com/invocations/default/56f92278-6e1e-481a-9d53-380ca7fb37ad?testReportRun=1&testReportShard=1&testReportAttempt=1#targets-Ly9wa2cvY3Jvc3NjbHVzdGVyL2xkcnJhbmRnZW46bGRycmFuZGdlbl90ZXN0) on master @ [a627560a261fdb778b5f818aad63013916f482c7](https://github.com/cockroachdb/cockroach/commits/a627560a261fdb778b5f818aad63013916f482c7):


```
=== RUN   TestGenerateLDRTable
    test_log_scope.go:171: test logs captured to: outputs.zip/logTestGenerateLDRTable1817933073
    test_log_scope.go:82: use -show-logs to present logs inline
    test_server_shim.go:154: automatically injected an external process virtual cluster under test; see comment at top of test_server_shim.go for details.
    logical_test.go:47: creating table: CREATE TABLE test_writer (col0_0 REGROLE NOT NULL, col0_1 INT2, col0_2 INTERVAL NOT NULL, col0_3 BPCHAR NOT NULL, col0_4 TSQUERY NOT NULL, col0_5 REFCURSOR[] NOT NULL, col0_6 UUID NULL, col0_7 DECIMAL NOT NULL, col0_8 TSVECTOR NOT NULL, col0_9 TIME NOT NULL, col0_10 REGTYPE, col0_11 TSQUERY NOT NULL, PRIMARY KEY (col0_0 DESC), INDEX (col0_2) STORING (col0_3, col0_7, col0_9, col0_10, col0_11))
    logical_test.go:57: error executing query="CREATE LOGICALLY REPLICATED TABLE b.test_writer FROM TABLE a.test_writer ON 'external://a' WITH BIDIRECTIONAL ON 'external://b'" args=[]: pq: cannot create logical replication stream: RefCursor is not supported by LDR
    panic.go:615: -- test log scope end --
test logs left over in: outputs.zip/logTestGenerateLDRTable1817933073
--- FAIL: TestGenerateLDRTable (1.58s)
```

Parameters:
 - <code>attempt=1</code>
 - <code>run=1</code>
 - <code>shard=1</code>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
/cc @cockroachdb/disaster-recovery
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestGenerateLDRTable.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/bazci/githubpost/issues)

</sub>


Jira issue: CRDB-56354

crosscluster/ldrrandgen.TestGenerateLDRTable [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_UnitTests/20738167?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_UnitTests/20738167?buildTab=artifacts#/) on master @ [930e2ef1de9b1220f52b6d15663b9e067778c76f](https://github.com/cockroachdb/cockroach/commits/930e2ef1de9b1220f52b6d15663b9e067778c76f):


```
=== RUN   TestGenerateLDRTable
    test_log_scope.go:171: test logs captured to: /artifacts/tmp/_tmp/1ee0ce66df5f0964c8d393c1246a3667/logTestGenerateLDRTable2213839036
    test_log_scope.go:82: use -show-logs to present logs inline
    logical_test.go:47: creating table: CREATE TABLE test_writer (col0_0 INTERVAL NOT NULL, col0_1 LTREE, col0_2 INT2 NOT NULL, col0_3 INT8 NOT NULL, col0_4 DECIMAL NULL, col0_5 REFCURSOR[] NOT NULL, col0_6 DATE NOT NULL, col0_7 TIMETZ NOT NULL, col0_8 VECTOR NOT NULL, col0_9 STRING AS (lower(CAST(col0_1 AS STRING))) VIRTUAL, col0_10 DECIMAL NULL AS (abs(col0_4)) STORED, col0_11 INT8 NOT NULL AS (col0_2 + col0_3) VIRTUAL, col0_12 INT8 NOT NULL AS (col0_2 + col0_3) VIRTUAL, col0_13 STRING NOT NULL AS (lower(CAST(col0_7 AS STRING))) VIRTUAL, col0_14 STRING NOT NULL AS (lower(CAST(col0_7 AS STRING))) STORED, col0_15 STRING NOT NULL AS (lower(CAST(col0_8 AS STRING))) STORED, PRIMARY KEY (col0_0, col0_3 DESC), INDEX (col0_0) STORING (col0_1, col0_5, col0_6, col0_8, col0_10))
    logical_test.go:57: error executing query="CREATE LOGICALLY REPLICATED TABLE b.test_writer FROM TABLE a.test_writer ON 'external://a' WITH BIDIRECTIONAL ON 'external://b'" args=[]: pq: cannot create logical replication stream: RefCursor is not supported by LDR
    panic.go:615: -- test log scope end --
test logs left over in: /artifacts/tmp/_tmp/1ee0ce66df5f0964c8d393c1246a3667/logTestGenerateLDRTable2213839036
--- FAIL: TestGenerateLDRTable (0.95s)
```
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestGenerateLDRTable.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/bazci/githubpost/issues)

</sub>

crosscluster/ldrrandgen.TestGenerateLDRTable [failed](https://mesolite.cluster.engflow.com/invocations/default/bec1578b-0172-4d99-9456-bdfd9043554b?testReportRun=10&testReportShard=1&testReportAttempt=1#targets-Ly9wa2cvY3Jvc3NjbHVzdGVyL2xkcnJhbmRnZW46bGRycmFuZGdlbl90ZXN0) on master @ [6816c43040b24bb221d00717c91f6a68a1a8e191](https://github.com/cockroachdb/cockroach/commits/6816c43040b24bb221d00717c91f6a68a1a8e191):


```
=== RUN   TestGenerateLDRTable
    test_log_scope.go:171: test logs captured to: outputs.zip/logTestGenerateLDRTable2935767202
    test_log_scope.go:82: use -show-logs to present logs inline
    test_server_shim.go:155: automatically injected a shared process virtual cluster under test; see comment at top of test_server_shim.go for details.
    logical_test.go:47: creating table: CREATE TABLE test_writer (col0_0 TSVECTOR NOT NULL, col0_1 VECTOR, col0_2 DATE NOT NULL, col0_3 TIMESTAMP NULL, col0_4 TIME, col0_5 REFCURSOR[] NULL, col0_6 STRING NOT NULL, col0_7 BPCHAR NOT NULL, col0_8 DECIMAL NOT NULL, col0_9 REGTYPE NULL, col0_10 VARCHAR NOT NULL, col0_11 BIT(43) NOT NULL, col0_12 REGTYPE NOT NULL, col0_13 STRING NOT NULL AS (lower(col0_7)) STORED, col0_14 STRING NULL AS (CASE WHEN col0_5 IS NULL THEN '8n-1$' ELSE '' END) VIRTUAL, PRIMARY KEY (col0_12))
    logical_test.go:57: error executing query="CREATE LOGICALLY REPLICATED TABLE b.test_writer FROM TABLE a.test_writer ON 'external://a' WITH BIDIRECTIONAL ON 'external://b'" args=[]: pq: cannot create logical replication stream: RefCursor is not supported by LDR
    panic.go:615: -- test log scope end --
test logs left over in: outputs.zip/logTestGenerateLDRTable2935767202
--- FAIL: TestGenerateLDRTable (1.52s)
```

Parameters:
 - <code>attempt=1</code>
 - <code>run=10</code>
 - <code>shard=1</code>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestGenerateLDRTable.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/bazci/githubpost/issues)

</sub>

crosscluster/ldrrandgen.TestGenerateLDRTable [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_UnitTests/20753373?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_UnitTests/20753373?buildTab=artifacts#/) on master @ [560b635c16f790db75a0fdc7492be9ecd8e96fa6](https://github.com/cockroachdb/cockroach/commits/560b635c16f790db75a0fdc7492be9ecd8e96fa6):


```
=== RUN   TestGenerateLDRTable
    test_log_scope.go:171: test logs captured to: /artifacts/tmp/_tmp/1ee0ce66df5f0964c8d393c1246a3667/logTestGenerateLDRTable4056361270
    test_log_scope.go:82: use -show-logs to present logs inline
    logical_test.go:47: creating table: CREATE TABLE test_writer (col0_0 INTERVAL NOT NULL, col0_1 STRING COLLATE da_DK NOT NULL, col0_2 UUID NULL, col0_3 TIMESTAMPTZ[] NULL, col0_4 INTERVAL NOT NULL, col0_5 STRING NOT NULL, col0_6 NAME NULL, col0_7 GEOGRAPHY, col0_8 TIMESTAMPTZ NULL, col0_9 OID, col0_10 INET NOT NULL, col0_11 REFCURSOR[] NOT NULL, col0_12 INTERVAL NOT NULL, col0_13 STRING NULL AS (CASE WHEN col0_8 IS NULL THEN e'\x07\x17\x0eQ_\x06' ELSE '' END) VIRTUAL, col0_14 STRING NOT NULL AS (lower(CAST(col0_10 AS STRING))) VIRTUAL, col0_15 STRING AS (lower(CAST(col0_9 AS STRING))) STORED, col0_16 STRING NULL AS (CASE WHEN col0_8 IS NULL THEN 'ysj7}' ELSE NULL END) STORED, col0_17 STRING AS (lower(CAST(col0_7 AS STRING))) STORED, PRIMARY KEY (col0_0 ASC))
    logical_test.go:57: error executing query="CREATE LOGICALLY REPLICATED TABLE b.test_writer FROM TABLE a.test_writer ON 'external://a' WITH BIDIRECTIONAL ON 'external://b'" args=[]: pq: cannot create logical replication stream: RefCursor is not supported by LDR
    panic.go:615: -- test log scope end --
test logs left over in: /artifacts/tmp/_tmp/1ee0ce66df5f0964c8d393c1246a3667/logTestGenerateLDRTable4056361270
--- FAIL: TestGenerateLDRTable (0.71s)
```
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestGenerateLDRTable.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/bazci/githubpost/issues)

</sub>

--- a/pkg/crosscluster/ldrrandgen/logical.go
+++ b/pkg/crosscluster/ldrrandgen/logical.go
@@ -15,6 +15,30 @@ import (
 	"github.com/cockroachdb/cockroach/pkg/sql/types"
 )
 
+func isSupportedType(t *types.T) bool {
+	switch t.Family() {
+	case types.RefCursorFamily:
+		// We don't support RefCursor columns in LDR tables because they do not
+		// support equality.
+		return false
+	case types.ArrayFamily:
+		// We don't allow Arrays of bits because rand.LoadTable doesn't correctly identify their bit width.
+		if t.ArrayContents().Family() == types.BitFamily {
+			return false
+		}
+		return isSupportedType(t.ArrayContents())
+	case types.TupleFamily:
+		for _, elemType := range t.TupleContents() {
+			if !isSupportedType(elemType) {
+				return false
+			}
+		}
+		return true
+	default:
+		return true
+	}
+}
+
 func GenerateLDRTable(
 	ctx context.Context, rng *rand.Rand, tableName string, supportKVWriter bool,
 ) *tree.CreateTable {
@@ -31,21 +55,7 @@ func GenerateLDRTable(
 		randgen.WithPrimaryIndexRequired(),
 		randgen.WithSkipColumnFamilyMutations(),
 		randgen.WithColumnFilter(func(columnDef *tree.ColumnTableDef) bool {
-			// We don't allow Arrays of bits because rand.LoadTable doesn't correctly identify their bit width.
-			if columnDef.Type.(*types.T).Family() == types.ArrayFamily && columnDef.Type.(*types.T).ArrayContents().Family() == types.BitFamily {
-				return false
-			}
-			// We don't support RefCursor columns in LDR tables because they do not
-			// support equality.
-			if columnDef.Type.(*types.T).Family() == types.RefCursorFamily {
-				return false
-			}
-			// We don't allow the special '"char"' column because pgwire truncates the value to 1 byte.
-			// TODO(jeffswenson): remove this once #149427 is fixed.
-			if columnDef.Type.(*types.T).Family() == types.StringFamily && columnDef.Type.(*types.T).Width() == 1 {
-				return false
-			}
-			return true
+			return isSupportedType(columnDef.Type.(*types.T))
 		}),
 		randgen.WithPrimaryIndexFilter(func(indexDef *tree.IndexTableDef, columnDefs []*tree.ColumnTableDef) bool {
 			for _, col := range indexDef.Columns {

